---
title: "Read the PACs data from Sierra"
author: "Xiaohui Wu, Xingyu Bi, Wenbin Ye"
date: "2023-09-20"
output:
  html_document: 
    toc: yes
vignette: >
  %\VignetteIndexEntry{Read_PAC_data_from_Sierra}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,warning = FALSE)
```


# Overview

This documentation describes how to use `Sierra` to detect poly(A) sites(or called poly(A) site clusters,PACs) from single-cell RNA-seq data, read the poly(A) sites data from `Sierra`, and then analyze it with `vizAPA`. We used  single-cell RNA sequencing data from peripheral blood mononuclear cells(PBMCs) for demonstration.The specific procedure for Sierra identification and quantification of poly(A) sites can be found in:[Sierra](https://github.com/VCCRI/Sierra/wiki/Sierra-Vignette)


# Read PACs data from Sierra
Identification of poly(A) sites from scRNA-seq data with Sierra generates three filesï¼š(1)`pbmc2_count/matrix.mtx.gz`(a sparse matrix where rows are the peaks and columns are the cells: records the expression of identified  peaks in a single cell); (2)`pbmc2_extract_peaks.txt`(records the coordinates information of identified peaks); (3)`pbmc2_peak_annotations.txt`(annotated peak imformation: records if peaks overlap exons, introns or UTRs). Then we can also convert PACs data from Sierra to PACdataset object for downstream analysis using vizAPA package.

```{r results='hide'}
library(vizAPA)
library(movAPA)
```

```{r eval=FALSE}
filepath <- "./PBMC/sierra/"
peakfile <- paste0(filepath,"pbmc2_peak_annotations.txt")
countfile <-paste0(filepath,"pbmc2_count")

```

```{r eval=FALSE}
peak_annotations <- read.delim(peakfile)
head(peak_annotations[1:3,])

#Read in peak data saved in MEX format. Files can be in a gzipped (.gz) format.
count <- Sierra::ReadPeakCounts(data.dir=countfile,mm.file="matrix.mtx.gz",barcodes.file ="barcodes.tsv.gz")
count <- as.data.frame(count)
head(count[1:3,1:4])

#Remove the numbers such as -1 in barcode
colnames(count)<-gsub("-1","",colnames(count))
```


```{r eval=FALSE}
#Keep the same peaks in count as in peak.annotation
infor <- intersect(row.names(count),row.names(peak_annotations))
peak_annotations<-peak_annotations[infor,]
count<-count[infor,]

peak_annotations$coord <- 0
#When strand="+", the end position of the identified peak is used as the coordinate of the pA site
peak_annotations[peak_annotations$strand == "+",]$coord <- peak_annotations[peak_annotations$strand == "+",]$end
#When strand="-",the starting position of the identified peak is used as the coordinate of the pA site
peak_annotations[peak_annotations$strand == "-",]$coord <- peak_annotations[peak_annotations$strand == "-",]$start

anno<- peak_annotations[,c("seqnames","start","end","strand","coord","gene_id")]
colnames(anno) <- c("chr","start","end","strand","coord","gene_id")
head(anno[1:3,1:6])

#Remove duplicate peak data based on the annotation information of peak
anno <- unique(anno)

#Preserve PA sites with gene name
anno<- anno[!is.na(anno$gene_id),]
anno$gene_id <- gsub("[.].*","",anno$gene_id)
count<-count[rownames(anno),]
```

Convert Sierra results into PACdataset object. Here, we will only read the PACs data annotated with `Sierra::AnnotatePeaksFromGTF` into the PACdataset. We can also combine the result of the annotated peak and peak count and then convert it to PACdataset object.

```{r eval=FALSE}
#Retention of cells with annotated cell type information
meta<-read.table(paste0(filepath,"sierra_cell.meta.txt"),sep = "\t",header = TRUE)
row.names(meta)<-meta$barcode
dim(count)
length(intersect(meta$barcode,colnames(count)))
count<- count[,intersect(meta$barcode,colnames(count))]
#combine the result of the annotated peak and peak count
pafile<-cbind(anno,count)
head(pafile[,1:8])

coldata<-data.frame(barcode=colnames(pafile)[7:ncol(pafile)],row.names =colnames(pafile)[7:ncol(pafile)] )
scPACds <- readPACds(pacFile = pafile, colDataFile = coldata,noIntergenic = FALSE)

```

##  Remove internal priming
```{r eval=FALSE}
library(BSgenome.Hsapiens.NCBI.GRCh38)
bsgenome<-BSgenome.Hsapiens.NCBI.GRCh38
#check the consistency of chromosome names between PACdataset and BSgenome
isChrConsistent(scPACds,bsgenome)

scPACds =removePACdsIP(scPACds , bsgenome, returnBoth=FALSE, up=-140, dn=10, conA=6, sepA=NA,chrCheck = FALSE)
#merge pA sites within 24nt of each other
scPACds<-mergePACds_v0(scPACds, d=24)

dim(scPACds@counts)  
```

##  Add the cell type annotation information

```{r eval=FALSE}
loc<- match(scPACds@colData$group,meta$barcode)
scPACds@colData$celltype<-meta$celltype[loc]
```

##  Load genome annotation to an annoHub
Load genome annotation from gtf file.
```{r}
annoSource=new("annoHub")
gff<-useGff("./PBMC/Homo_sapiens.GRCh38.109.gtf.gz")
annoSource=addAnno(annoSource, gff)
```

##  Annotation 
After read the data into a PACdataset ,we annotate the PACs data using the gtf file(GRCh38). Please make sure the chromosome name of your PACs data is the same as the gtf file.

```{r eval=FALSE}
#check the consistency of chromosome names between PACdataset and genome annotations
isChrConsistent(scPACds,gff)
scPACds<-annotatePAC(scPACds,aGFF=gff)

# extend 3UTR by 1000bp
scPACds=ext3UTRPACds(scPACds, 1000)
save(scPACds,file =paste0(filepath,"pbmc2_sierra_APA.rda"))
```

The PAC data generated by Sierra can be downloaded by clicking on this link:[pbmc2_sierra_APA.rda](https://github.com/BMILAB/vizAPA/tree/refer)
```{r}
#Load annotated PAC data
load("./PBMC/sierra/pbmc2_sierra_APA.rda")
summary(scPACds)
```

Plots single nucleotide profile using the sequences of 3'UTR.
```{r eval=FALSE}
fafiles = faFromPACds(scPACds, bsgenome, what = "updn", fapre = "pbmc2.sierra.200nt.GRCh38",
                       up = -100, dn = 100, byGrp = "ftr", chrCheck = FALSE)

plotATCGforFAfile("pbmc2.sierra.200nt.GRCh38.3UTR.fa", ofreq = FALSE, opdf = TRUE, refPos = 101,
                  filepre = "")
```

```{r out.height="70%",out.width="60%",echo=FALSE}
knitr::include_graphics("C:/vizAPA/vizAPA data/PBMC/sierra/pbmc2.sierra.200nt.GRCh38.3UTR.png")
```

# Visualziation 
##  Cell cluster visualziation

Use `Seurat::reduceDim function` to get embeddings based on the counts data. Then plot the UMAP plot showing cell clusters and another UMAP plot overlaying with the mean expression value of poly(A) site in each cell.
```{r results='hide',out.height="90%",out.width="90%"}
## get embeddings using the pA count matrix with normalization
scPACds=reduceDim(scPACds, nfeature=2000, dims=1:10, dimLabel='umap_norm', norm=TRUE)
vizUMAP(scPACds, group='celltype', xcol='umap_norm1', ycol='umap_norm2' )

```

We selected 3'UTR APA genes and showed the expression level of  the gene with high expression pA sites across cell clusters.
```{r}
APAds=get3UTRAPAds(scPACds)
top10=APAds@anno[order(Matrix::rowSums(APAds@counts), decreasing = TRUE)[1:10],]
gene=top10$gene[1]
gene
```

```{r results='hide'}

statTheme=setStatTheme(list(scale.high.col="purple",scale.low.col="grey"),check = TRUE)
vizUMAP(scPACds, group='celltype', xcol='umap_norm1', ycol='umap_norm2' ,genes = gene,statTheme = statTheme)
```

##  APA markers visualization
We first calculate the RUD index for each gene and compare one cell type(CD4+Tcell) with all other cells to obtain APA markers.When `logFC = 0.25`(The minimum log2 multiple of the average expression of genes in a cell cluster relative to the average expression of all other cell clusters) was set, the difference in the usage of the distal pA sites between CD4+Tcell and other cells was not significant.
```{r results='hide'}
iPACds=getAPAindexPACds(scPACds, choose2PA='PD')
markers=getAPAmarkers(iPACds,  group='celltype',cluster1 = "CD4+Tcell" ,logFC = 0.25)
```

```{r}
table(markers$cluster1, markers$cluster2)
```


Next, the distal pA site usage was compared between Dendritic cells with higher average expression of pA sites and other cells to find APA markers.
```{r results='hide'}
markers=getAPAmarkers(iPACds,  group='celltype',cluster1 = "Dendritic cell" ,logFC = 0.25)

```

```{r }
table(markers$cluster1, markers$cluster2)
```


Visualisation of the top5 APA markers using bubble plot.
```{r}
vizAPAMarkers(iPACds, group='celltype', markers=markers$rowid[1:5], figType = 'bubble', statTheme=list(xgroup=FALSE))
```

We can also visualise APA markers with violin plot.
```{r}
vizAPAMarkers(iPACds, group='celltype', markers=markers$rowid[1:5], figType = 'violin')
```
