---
title: 'Using vizTracks in vizAPA: a full tutorial'
author: "Xiaohui Wu, Xingyu Bi, Wenbin Ye"
date: "2023-09-20"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Using_vizAPA_vizTracks_a_full_tutorial}
  %\VignetteEncoding{UTF-8}   
  %\VignetteEngine{knitr::rmarkdown}
---


# Overview

This tutorial takes a `PACdataset` object storing a list of poly(A) sites and BAM files as input and describes full usages of series function related to `vizTracks` in `vizAPA`.

# Data preparation

## Demo PACdataset

In the package of `vizAPA`, there is a demo `PACdataset` object of mouse sperm cells, containing 974 pAs [poly(A) sites] from 413 genes. There are total 955 cells from three cell types (SC, Spermatocytes; RS, Round spermatids; ES, Elongating spermatids). This PACdataset has been annotated, with both pAs' and cells' meta data.

```{r readPACds, eval=TRUE, message=FALSE, warning=FALSE}
library(vizAPA)

data(scPACds, package='vizAPA')

# summary of the PACdataset
movAPA::summary(scPACds)

# cell meta data
head(scPACds@colData)
```
It is better to change the level order of the factor `celltype` in scPACds, which can make the labels in the plots consistent with sperm differentiation from SC to RS to ES.

```{r celltype_order, message=FALSE, warning=FALSE}
scPACds@colData$celltype=factor(scPACds@colData$celltype, 
                                levels=c('SC','RS','ES'))
```


The raw data was obtained from NCBI (accession number: GSM280334), and pAs were extracted by [scAPAtrap](https://github.com/BMILAB/scAPAtrap/). The resulted pA list from scAPAtrap was then processed and annotated with MM10 genome annotation (TxDb.Mmusculus.UCSC.mm10.knownGene) in [movAPA](https://github.com/BMILAB/movAPA/), and a `PACdataset` object was finally obtained. This object was then used in vizAPA for visualization.

For demonstration Only genes containing at least one top 500 pAs ranked by the total number of reads were retained in the PACdataset.

Generally, it is easy to use `movAPA` to import the result (basically a pA/peak list with per cell counts) from `scAPAtrap` or other tools, following the example code below. Please see the vignettes "Read_PAC_data_from_Sierra" and "Read_PACs_data_from_scAPAtrap" for more details.

```{r notrun, eval=FALSE}
# First, use readPACds or createPACdataset to create the PACdataset object
# from a pA list, pA-cell counts table, cell meta data
PACds=movAPA::readPACds(pacFile,
                colDataFile)

movAPA::summary(PACds)

# If there are internal priming artifacts, should be removed first
library(BSgenome.Mmusculus.UCSC.mm10)
bsgenome=BSgenome.Mmusculus.UCSC.mm10
PACds=movAPA::removePACdsIP(PACds, bsgenome)
PACds=PACds$real

# Then annotate the pA list with genome annotation from TxDb or a GFF/GTF file
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb=TxDb.Mmusculus.UCSC.mm10.knownGene

PACds=movAPA::annotatePAC(PACds, aGFF=txdb)

# Then extend 3'UTR by some length, e.g., 1000 bp
# which can recuite potential pAs in downstream intergenic regions of 3'UTR
PACds=movAPA::ext3UTRPACds(PACds, 1000)

# For single cell data, we'd better only use 3'UTR pAs for analysis
# since pAs in other genomic regions may be artifacts
PACds=movAPA::get3UTRAPAds(PACds)

# Finally, a full annotated 3'UTR PACdataset can be obtained
# which could be used in vizAPA for visualization
movAPA::summary(PACds)
```

## BAM files

For this tutorial, three very small BAM files corresponding to the three cell types (SC, ES, and RS), which contains only mapped reads from five genes, were made from the original BAM files.

The `readBAMFileNames` function reads the BAM file names into a data frame recording file name, group, and label. Each row in the data frame stores the information of a BAM file. Normally, a BAM file represents a condition, e.g., cell type. The `readBAMFileNames` function will check the existence of each file (and the corresponding .bai file)! The BAM files could be in different folders, please see `?readBAMFileNames` for more examples.

The BAM files for the following analysis can be downloaded here:[mouse.sperm.bam](https://github.com/BMILAB/vizAPA/tree/refer)


```{r readBam}
#Create the list of BAM files
bam.files=c("dedup_GSM2803334.ES.mini.sorted.bam",
            "dedup_GSM2803334.RS.mini.sorted.bam",
            "dedup_GSM2803334.SC.mini.sorted.bam")
bam.groups=c("ES","RS","SC")
bam.labels=c("ES" ,"RS","SC")
bam.path='./'

# here we set bam.order to make BAM colors consistent with PACdataset
bams<-readBAMFileNames(bam.files=bam.files, 
                       bam.path=bam.path, 
                       bam.labels = bam.labels, 
                       bam.groups = bam.groups,
                       bam.order=c('SC','RS','ES'))
bams
```

# Genome annotation

In vizAPA, the genome annotation is used for the track plots to show gene models in a genomic region. The genome annotation could be retrieved from several sources, including gff3/gtf file, TxDb, EnsDb, BioMart, and OrganismDb. Users can provide one or more annotation sources. Even if no publicly availabe genome annotation exists, vizAPA can plot simply a gene or a pA region to represent the gene model.

Following are commonly used genome annotation sources:

-   OrganismDb object: recommended, support gene symbols and other combination of columns as label.

-   TxDb object: don't support gene symbol labeling.

-   EnsDb object: supports gene symbol labeling, filtering etc.

| Object type   | example package/object                                    | name contents                          |
|------------------|-------------------------|-----------------------------|
| OrgDb         | org.Hs.eg.db                                       | gene based info. for Homo sapiens      |
| TxDb          | TxDb.Hsapiens.UCSC.hg19.knownGene                  | transcriptome ranges for Homo sapiens  |
| OrganismDb    | Homo.sapiens                                       | composite information for Homo sapiens |
| gff3/gtf file | Homo_sapiens.GRCh38.96.gtf                         | gtf file for Homo sapiens              |
| biomaRt       | hsapiens_gene_ensembl                              | biomaRt for Homo sapiens               |
| genes         | a data frame with chr/strand/start/end/gene_xx_ids | customized annotation in vizAPA        |

We can make an `annoHub` object with elements named txdb/gff/orgdb/ensdb/biomart/genes in its `annos` slot denoting different annotation sources. This `annoHub` object could be used by many functions in vizAPA.

```{r annoHub}
annoSource=new("annoHub")
```

## Using a GFF3/GTF file

Given a GFF3/GTF file, the `useGff` function can be used to parse the GFF3/GTF file to a data frame, which can be added to the annoHub. Here is the example code. However, it would be more convenient to use other sources of genome annotations, e.g., EnsDb, BioMart, and OrganismDb,

```{r addAnno_gff_notrun, eval=FALSE}
gff <- useGff(gff ="gencode.vM32.annotation.gff3")
annoSource=addAnno(annoSource, gff)
```

## Using OrganismDb

We recommend to use `OrganismDb` as it contains both gene symbol and entrez id, which is easier to use.

```{r addAnno_orgdb, results='hide', warning=FALSE, message=FALSE}
library(Mus.musculus, quietly = TRUE)
orgdb=Mus.musculus
annoSource=addAnno(annoSource, orgdb)
annoSource=setDefaultAnno(annoSource, 'orgdb')
```

## Using TxDb

The annoHub object in vizAPA allows adding multiple annotations, users can set default annotation for visualization.

```{r addAnno_txdb, warning=FALSE, message=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene, quietly = TRUE)
txdb=TxDb.Mmusculus.UCSC.mm10.knownGene
annoSource=addAnno(annoSource, txdb)
```

## Using EnsDb

```{r addAnno_ensdb, warning=FALSE, message=FALSE}
library(EnsDb.Mmusculus.v79, quietly = TRUE)
ensdb=EnsDb.Mmusculus.v79
annoSource=addAnno(annoSource, ensdb)
```

## Using BioMart

```{r addAnno_biomart, warning=FALSE, message=FALSE}
library(biomaRt)
bm = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
annoSource=addAnno(annoSource, bm)
```

## Using customized genome annotation

The `annoHub` supports user provided data frame with genomic ranges of genes, which is named `genes` in `\@annos`. Here we extracted all genes from `OrganismDb` and add it to `annoSource` for demonstration.

```{r addAnno_genes}
genes=getAnnoGenes(orgdb)
annoSource=addAnno(annoSource, genes)
```

## Set priority of annotations

When searching given genes for genomic ranges in an `annoHub`, the searching order is the default annotation followed by other annotations in the order of `\@annos`' elements. We can set orders of the priority by reorder the elements in the annos slot.

```{r annoHub_order}
## re-order annoSource to set the searching priority
names(annoSource@annos)
annoSource@annos=annoSource@annos[c('orgdb','txdb','ensdb','genes','biomart')]
annoSource
```

## Chromosome consistency

If multiple sources of genome annotation are provided, we can check the consistency of chromosome names among different sources, use the function `isChrConsistent`. Normally, main chromosome names (e.g., chr1, 1) are the same across different annotations, scaffolds may be different, we can set `exact=FALSE` to allow partial overlapping of chromosome names.

```{r check_chrs, warning=FALSE, message=FALSE}
## The chr names are different: orgdb is chr1, ensdb is 1.
isChrConsistent(annoSource['orgdb'], annoSource['ensdb'], exact=TRUE) 

isChrConsistent(annoSource['orgdb'], annoSource['ensdb'], exact=FALSE) 

## The chr names are total the same between txdb and orgdb, both are like chr1
isChrConsistent(annoSource['txdb'], annoSource['orgdb'], exact=TRUE) 
```

We can also check the consistency of chromosome names among PACdataset, BAM files, and genome annotations.

```{r getChrs, warning=FALSE, message=FALSE}
# chr names are the same in PACds and in BAM file
isChrConsistent(scPACds, bams, exact=TRUE)

# Here for annoSource, the default anno was used. 
# It seems that some chromosomes do not have exacytly the same name
isChrConsistent(scPACds, bams, annoSource, exact=TRUE) 
# However, the main chr names are the same
isChrConsistent(scPACds, bams, annoSource, exact=FALSE) 

# We then get full chromosome names of each object, 
# and it would be fine that main chromosomes have the same name.
getChrs(scPACds)
getChrs(bams)
```

We can also get chr names for different annotations in `annoSource`.
```{r getChrs_annos, eval=FALSE}
getChrs(annoSource, which='orgdb') # chr1
getChrs(annoSource, which='ensdb') # 1
getChrs(annoSource, which='biomart') # 1
getChrs(annoSource, which='genes') # chr1
getChrs(annoSource, which='txdb') # chr1
```

## Chromosome name mapping

Since there are different annotations in the `annoSource`, we can add the `chrMappings` slot to coordinate the chromosome names among different annotation sources. Given `chrMappings`, vizAPA can automatically use appropriate chr names in its functions.

```{r chr_mapping}
chrMappings=data.frame(cn1=c(1:19, 'X','Y'))
chrMappings$cn2=paste0('chr', chrMappings$cn1)
annoSource@chrMappings=chrMappings
head(chrMappings)
```

# `vizTracks` to plot gene model, pAs and BAM tracks

`vizTracks` gets all tracks for a genomic region, a gene, or PAs, including gene model track, pA track(s), cells track, and BAM track(s).

-   **gene model track**: plot gene models given a specific region or a gene symbol/ID, according to the annotation(s) in the `annoHub` object.
-   **pA track**: plot the positions, expression levels, APA ratios for a PACdataset or multiple PACdataset objects.
-   **cells track**: plot counts or ratios of individual cells using gradient colors.
-   **BAM track**: plot BAM coverage using lines or areas for individual cell groups or merged cell groups.

## Plot tracks in a gene

Here we plot for a gene all available tracks including gene model, BAM coverages, pA coordinates, and individual cell distributions. The gene model is based on the annotation provided in `annoSource` (here is the default element `annoSource['orgdb']`). The pAs in this gene are represented as a expanded 21 bp (2\*`PA.width`+1) region of `coord`. Only the position rather than the expression of pAs is shown (PA.show='pos'). BAM coverage information is from `bams`. The 5' and 3' ranges of the gene model are expanded by 300 bp (`space5` and `space3`).

### Gene information

In the following we used gene Asrgl1 (Entrezid=66514) for demonstration, as this gene was present in our demo BAM files and in demo PACdataset.

```{r select_APA}
gene=66514

## show this gene in PACds
scPACds@anno[scPACds@anno$gene==gene, c(1:6, 10:12)]

## the gene is represented as Entrez ID in PACds, we can show its symbol name 
## using 'genes' table in the annos slot
annoSource@annos$genes[annoSource@annos$genes$gene_entrezid==gene, ]
```

### Basic `vizTracks` for a gene

Plot an example gene, with gene model, pA coordinates, and BAM coverages. Here we extend the gene model by 1000 bp at both ends.

```{r vizTracks1, fig.show="hold", message=FALSE, warning=FALSE}
# here gm.reduce=FALSE to show full features of gene model instead of collapsed
vizTracks(gene=gene, 
          PACds.list=list(pA=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000,
          vizTheme=list(gm.reduce=FALSE))

# plot collapsed gene model and the BAM coverage
vizTracks(gene=gene, 
          bams=bams, PACds.list=list(pA=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000)

```

### Add heatmap for individual cells

We can also plot the pA expression in individual cells as a heatmap.

```{r vizTracks1_cells, message=FALSE, warning=FALSE}
vizTracks(gene=gene, 
          bams=bams, PACds.list=list(pA=scPACds), PA.show=c("pos"),
          cells=TRUE, cells.group='celltype',
          cells.method=c('sum'), cells.sort=c('group'),
          cells.width=100,
          annoSource=annoSource,
          PA.columns="coord", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000)

```

### Change colors by `vizTheme`

However, in the above figure, the colors are not consistent between the BAM track and the cells track, as different track are independent. We can set `vizTheme` to make cells' group colors as the order or BAM files. The global variable `vizTHEME` stores all visualization parameters for vizAPA, see `?setVizTheme` for details.

```{r vizTheme, eval=FALSE}
# Show default vizTheme parameters
setVizTheme(NULL)
```

```{r vizTracks1_colors, message=FALSE, warning=FALSE }
# set colors of the cells track and BAM coverage
cells.group.cols=c('red','green','blue')
names(cells.group.cols)=c('ES','RS','SC')
vizTheme=list(cells.group.cols=cells.group.cols, bams.col=cells.group.cols)

vizTracks(gene=gene, 
          bams=bams, PACds.list=list(pA=scPACds), PA.show=c("pos"),
          cells=TRUE, cells.group='celltype',
          cells.method=c('sum'), cells.sort=c('group'),
          cells.width=100,
          annoSource=annoSource,
          PA.columns="coord", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000,
          vizTheme=vizTheme)
```

## Plot tracks in a specified genomic region

In addition to plot tracks in a given gene, we can also plot tracks in a given genomic region by setting `genomicRegion` instead of `gene` in `vizTracks`.

### Get gene's region

```{r get_gene_gr, message=FALSE, warning=FALSE}
# here for demonstration, we get the genomic region of the gene
genesGR=getGenesRange(gene, annoSource, rt='str')
genesGR
```

### Basic `vizTracks` for a region

```{r vizTracks2, warning=FALSE, message=FALSE}
vizTracks(genomicRegion=genesGR, 
          bams=bams, PACds.list=list(pA=scPACds), PA.show=c("pos"),
          cells=TRUE, cells.group='celltype',
          cells.method=c('sum'), cells.sort=c('group'),
          cells.width=100,
          annoSource=annoSource,
          PA.columns="coord", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000,
          vizTheme=NULL)
```

## Customize pA tracks

This demo `PACdataset` provides also the range of each pA (the start and end columns in `PACds\@anno`, so we can represent pAs as a range by providing the `PA.columns="start:end"`. In this case, the value of `PA.width` or `cells.width` will be ignored, and the bar width of each pA is the `start-end+1` of the respective pA.

### Plot pA regions

```{r vizTracks_PA_region, fig.show="hold", message=FALSE, warning=FALSE}
# show the region of the pA
scPACds@anno[1:5, 1:5]

# plot pA regions by setting PA.columns
vizTracks(genomicRegion=genesGR, 
          bams=bams, PACds.list=list(pA=scPACds), PA.show=c("pos"),
          cells=FALSE,
          annoSource=annoSource,
          PA.columns="start:end", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000,
          vizTheme=NULL)
```

### Plot pA values

We can also add additional tracks of expression values (counts of ratios) of pAs, by adding the column(s) of expression value in `PACds\@anno` or `PACds\@counts` to `PA.show`.

```{r vizTracks_PA_value, fig.show="hold", message=FALSE, warning=FALSE}
## Here we add a tot_tagnum to the PACds@anno, 
## and show the total expression of pAs
scPACds@anno$tot_tagnum=Matrix::rowSums(scPACds@counts)

## We can do log-transformation to show the tot_tagnum clearly.
scPACds@anno$tot_tagnum=log2(scPACds@anno$tot_tagnum+1)

vizTracks(genomicRegion=genesGR, 
          bams=NULL, PACds.list=list(pA=scPACds), 
          PA.show=c("pos", "tot_tagnum"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000,
          vizTheme=NULL)
```

### Plot multiple pA datasets

If we have multiple PACdatasets (e.g., one for bulk and one for single cell), we can plot coordinates and expression values of each PACdataset as individual tracks by adding the PACdataset in `PACds.list`. In the following example, for each PACdataset, both the coordinate and the expression value of each PACdataset will be shown.

```{r vizTracks_multi_PA, fig.show='asis', message=FALSE, warning=FALSE}

# here we just use a replicated PACds for demonstration
# we can save the tracks and then plot
tks=vizTracks(genomicRegion=genesGR, 
          bams=NULL, PACds.list=list(pA1=scPACds, PA2=scPACds), 
          PA.show=c("pos", "tot_tagnum"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10, logPA=TRUE,
          space5=1000, space3=1000,
          vizTheme=NULL, res='list')

# plot tracks
ggbio::tracks(tks)

# we can change the title of the tracks
names(tks)[2:5]=c('pos1', 'tag1', 'pos2', 'tag2')
ggbio::tracks(tks)
```

### Change colors by `vizTheme`

The `vizTHEME` variable contains all settings for the track plot, which can be customized. Here we changed the colors of pAs.

```{r vizTracks_PA_colors, fig.show="hold", message=FALSE, warning=FALSE}
vizTheme=list(PA.col='green', PA.fill='green')
vizTracks(gene=gene,
          bams=NULL, PACds.list=list(pA1=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000,
          vizTheme=vizTheme)
```

## Customize the cells track

Once set `cells=TRUE` in `vizTracks`, a track would be plot to show pA counts or ratios in individual cells. The bar width of each pA depends on `PA.columns` and/or `PA.width`. If `PA.columns` is like 'coord' then the bar width is 2\*PA.width+1; if `PA.columns` is like 'start:end', then the bar width of a pA is `end-start+1` of the respective pA.

By default, the cells are sorted within the group, but we can also do not sort the cells or sort the cells within all cells. We can also customize the colors of the values and cell annotations.

### Plot expression levels for each cell

```{r vizTracks_cells_1, fig.show='asis', message=FALSE, warning=FALSE}
# use diffusion map to sort cells within each group
vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='diff', cells.sort='group', 
          cells.width=100,
          PA.columns="coord", PA.width=10)
```

### Plot log2 level

The pA expression of one pA is too large to show clearly the expression values for other pAs, here we log2 the value. We can also sort cells by the total counts or ratios of each cell.

```{r vizTracks_cells_2, fig.show='asis', message=FALSE, warning=FALSE}
vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='sum', cells.sort='group', 
          cells.width=100,
          PA.columns="coord", PA.width=10, logPA=TRUE)
```

### Change cell annotation bar

We can change the position of the cell annotation bar to the right, and also change the colors of the value bar.

```{r vizTracks_cells_3, fig.show='asis', message=FALSE, warning=FALSE}
vizTheme=list(cells.annoBar.pos='right', 
              cells.scale.low='white', 
              cells.scale.high='purple', cells.scale.mid = "grey",
              cells.group.cols=c(RColorBrewer::brewer.pal(8, "Set1")))

vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='sum', cells.sort='group', 
          cells.width=100,
          PA.columns="coord", PA.width=10, logPA=TRUE, vizTheme=vizTheme)
```

We can also put the annotation bar in a specific position.

```{r vizTracks_cells_4, fig.show='asis', message=FALSE, warning=FALSE}
vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='sum', cells.sort='group', 
          cells.width=100, cells.annoCoord=9110000,
          PA.columns="coord", PA.width=10, logPA=TRUE)
```

### Sort cells

By default, the cells are sorted within the group, we can also do not sort the cells.

```{r vizTracks_cells_5, fig.show='asis', message=FALSE, warning=FALSE}
vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='sum', cells.sort='none', 
          cells.width=100, 
          PA.columns="coord", PA.width=10, logPA=TRUE)
```

We can also sort the cells within all cells rather than within each group.

```{r vizTracks_cells_6, fig.show='asis', message=FALSE, warning=FALSE}
vizTracks(genomicRegion=genesGR, 
          PACds.list=scPACds, PA.show=c("pos"),
          annoSource=annoSource,
          cells=TRUE, cells.group='celltype', 
          cells.method='sum', cells.sort='all', 
          cells.width=100, 
          PA.columns="coord", PA.width=10, logPA=TRUE)

```

## Customize BAM tracks

### Plot BAM coverage

By default, BAM coverage of individual BAM files listed in `bams` will be plot. Each BAM could be one cell type. We can also add additional track showing merged (average or sum) coverage from all BAM files, by setting `vizTheme`.

```{r vizTracks_BAM1, fig.show="hold", message=FALSE, warning=FALSE}
# set how to show BAM coverage
vizTheme=list(bam.covMerge='avg', 
              bam.showMerge=TRUE, 
              bam.showSingle=TRUE, 
              bam.covMerge.col='yellow', 
              bam.covMerge.fill='yellow')

# we can also set the colors for individual bam coverages
vizTheme$bams.col=RColorBrewer::brewer.pal(8, "Set1")
vizTheme$bams.fill=RColorBrewer::brewer.pal(8, "Set1")

vizTracks(gene=gene,
          bams=bams, 
          PACds.list=list(pA1=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000,
          vizTheme=vizTheme)
```

### Collapse BAM coverage

We can also collapse all BAM coverage into one track by setting `bam.collapse` in `vizTheme` as TRUE. In this case, BAM coverage of individual BAM files will be shown as curves in one track.

```{r vizTracks_BAM2, fig.show="hold", message=FALSE, warning=FALSE}
vizTheme$bam.collapse=TRUE

vizTracks(gene=gene,
          bams=bams, 
          PACds.list=list(pA1=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000,
          vizTheme=vizTheme)
```

## Customize the gene model track

### Plot a gene model

vizAPA can retrieve gene model track from a given gene or a genomic range, based on different annotation sources, by setting `annoSource`. Here is an example to get gene model from a TxDb object. Before start, we check whether the chromosome names in this TxDb object are consistent with PACdataset and BAM files.

```{r check_chrs2, message=FALSE, warning=FALSE}
## yes, they are consistent
isChrConsistent(annoSource['txdb'], bams, scPACds, exact=FALSE) 
head(getChrs(annoSource['txdb']))

## set the default anno as txdb
annoSource=setDefaultAnno(annoSource, 'txdb')
```

```{r vizTracks_gene1, message=FALSE, warning=FALSE}
vizTracks(gene=gene, 
          bams=bams, 
          PACds.list=list(pA=scPACds), PA.show=c("pos"),
          annoSource=annoSource,
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000)
```

### Plot a genomic region

We can also plot the genomic region instead of gene model, without providing any annotations, i.e., just providing an empty annoSource.

```{r vizTracks_gene2, message=FALSE, warning=FALSE}
vizTracks(gene=gene, 
          bams=bams, 
          PACds.list=list(pA=scPACds), PA.show=c("pos"),
          annoSource=new('annoHub'),
          PA.columns="coord", PA.width=10,
          space5=1000, space3=1000)

```

## Plot tracks mannually

The `vizTracks` function provide a wrapper function to plot BAMs, pAs, gene model, individual cells together. However, we can also use a series of `getTrack...` functions to get gene model tracks from different genome annotation sources, get pA tracks from different PACdatasets, get pA in individual cells, and get BAM coverage tracks from different BAM files, respectively. And then we can combine all these tracks together in one plot.

### Get gene model tracks

```{r get_gene_grs, eval=FALSE}
# first check whether the gene is annotated in different annos.
getGenesRange(genes=gene, annoSource['txdb'], rt='gr')
getGenesRange(genes=gene, annoSource['ensdb'], rt='gr')
getGenesRange(genes=gene, annoSource['orgdb'], rt='gr')
getGenesRange(genes=gene, annoSource['biomart'], rt='gr')
getGenesRange(genes=gene, annoSource['genes'], rt='gr')
```

```{r get_indiv_tracks_gm, message=FALSE, warning=FALSE}
# collapose gene model features, not show transcript labels
vizTheme=list(gm.reduce=FALSE, gm.label=FALSE)

# use the orgdb anno
annoSource=setDefaultAnno(annoSource, 'orgdb')
gm.tk1=getTrackGeneModel(gene=gene, annoSource=annoSource, title='orgdb',
                         vizTheme = vizTheme)

annoSource=setDefaultAnno(annoSource, 'ensdb') 
gm.tk2=getTrackGeneModel(gene=gene, annoSource=annoSource, 
                         title='ensdb', vizTheme = vizTheme)

annoSource=setDefaultAnno(annoSource, 'txdb') 
gm.tk3=getTrackGeneModel(gene=gene, annoSource=annoSource, 
                         title='txdb', vizTheme = vizTheme) 

annoSource=setDefaultAnno(annoSource, 'genes') 
gm.tk4=getTrackGeneModel(gene=gene, annoSource=annoSource, 
                         title='genes', vizTheme = vizTheme) 

#annoSource=setDefaultAnno(annoSource, 'biomart') 
#gm.tk5=getTrackGeneModel(gene=gene, annoSource=annoSource, title='biomart') 

```

```{r plot_indiv_tracks_gm, message=FALSE, warning=FALSE}
# show these different annotations for the same gene
tks=c(gm.tk1, gm.tk2, gm.tk3, gm.tk4)
ggbio::tracks(tks, heights=c(2, 3, 3, 1.5))
```

### Get pA/cells/BAM tracks

We can also get pA or cells tracks for PACdatasets.

```{r get_PA_tracks, message=FALSE, warning=FALSE}
pac.tk1=getTrackPACds(scPACds, gene=gene, 
                      PA.show=c("pos"), 
                      PA.columns="coord", PA.width=10,
                      title='PApos', vizTheme = list(PA.shape=17)) 

pac.tk2=getTrackPACds(scPACds, gene=gene,
                      PA.show=c("tot_tagnum"), log=TRUE,
                      PA.columns="start:end", title='totTag')

sc.tk1=getTrackCells(scPACds,gene=gene,
                     group='celltype', 
                     sortMethod='sum', sortCells='group',log=TRUE,
                     PA.columns="coord", PA.width=100, title='diffs')

genomicRegion=getGenesRange(genes=gene, annoObj=annoSource['orgdb'], rt='str')
bam.tk1=getTrackBams(bams, genomicRegion=genomicRegion) 

## show gene model, pAs, cells, and bams
tks=c(gm.tk1, pac.tk1, pac.tk2, sc.tk1 ,bam.tk1)
ggbio::tracks(tks)

```
We can zoom in on a genomic region by specifying `xlim` and set the height of each track. For example, here we only display the 3'UTR area.

```{r tracks_zoom, warning=FALSE, message=FALSE}
# plot legend ‘horizontal’ or ‘vertical’
ggbio::tracks(tks, heights=c(0.8, 0.5, 1.2, 1.2, 1.2, 1.2, 1.2),
              xlim=c(9111000, 9113500))+ 
  ggplot2::theme(legend.direction = 'vertical')
```

# Session information

The session information records the versions of all the packages used in the generation of the present document.

```{r sessionInfo}
sessionInfo()
```
